hooks/useApproveExchange.ts

```diff
@@ -0,0 +1,127 @@
+import useSynthetixQueries from '@synthetixio/queries';
+import { wei } from '@synthetixio/wei';
+import { ethers } from 'ethers';
+import React from 'react';
+import { useRecoilState, useRecoilValue, useSetRecoilState } from 'recoil';
+
+import { SYNTH_SWAP_OPTIMISM_ADDRESS } from 'constants/address';
+import Connector from 'containers/Connector';
+import TransactionNotifier from 'containers/TransactionNotifier';
+import { useExchangeContext } from 'contexts/ExchangeContext';
+import use1InchApproveAddressQuery from 'queries/1inch/use1InchApproveAddressQuery';
+import {
+	isApprovedState,
+	isApprovingState,
+	quoteCurrencyAmountState,
+	quoteCurrencyKeyState,
+	txErrorState,
+} from 'store/exchange';
+import logError from 'utils/logError';
+
+const useApproveExchange = () => {
+	const [isApproving, setIsApproving] = useRecoilState(isApprovingState);
+	const [isApproved, setIsApproved] = useRecoilState(isApprovedState);
+	const quoteCurrencyAmount = useRecoilValue(quoteCurrencyAmountState);
+	const setTxError = useSetRecoilState(txErrorState);
+
+	const { isWalletConnected, walletAddress } = Connector.useContainer();
+
+	const { useContractTxn } = useSynthetixQueries();
+	const { monitorTransaction } = TransactionNotifier.useContainer();
+	const quoteCurrencyKey = useRecoilValue(quoteCurrencyKeyState);
+	const {
+		txProvider,
+		quoteCurrencyContract,
+		needsApproval,
+		setOpenModal,
+		allTokensMap,
+	} = useExchangeContext();
+
+	const oneInchApproveAddressQuery = use1InchApproveAddressQuery({
+		enabled: txProvider === '1inch',
+	});
+
+	const oneInchApproveAddress = oneInchApproveAddressQuery.data ?? null;
+
+	const approveAddress =
+		txProvider === '1inch' ? oneInchApproveAddress : SYNTH_SWAP_OPTIMISM_ADDRESS;
+
+	const approveTxn = useContractTxn(
+		quoteCurrencyContract,
+		'approve',
+		[approveAddress, ethers.constants.MaxUint256],
+		undefined,
+		{ enabled: !!approveAddress && !!quoteCurrencyKey && needsApproval }
+	);
+
+	React.useEffect(() => {
+		if (approveTxn.hash) {
+			monitorTransaction({
+				txHash: approveTxn.hash,
+				onTxConfirmed: () => {
+					setIsApproving(false);
+					setIsApproved(true);
+				},
+			});
+		}
+
+		// eslint-disable-next-line
+	}, [approveTxn.hash]);
+
+	const handleApprove = async () => {
```

Comment:
Should probably wrap with useCallback in case it's used a callback / memo dep somewhere else

Reply:
Fair enough. Will fix.